tests/test_file_upload.ts

```diff
@@ -0,0 +1,346 @@
+            target: null
+        };
+
+        expect(() => file_upload_component.add_files_from_button(mock_event)).toThrow(
+            'Target is null'
+        );
+    });
+
+    test('add_files_from_button throws an error when event.target.files is null', () => {
+        // tslint:disable-next-line:no-object-literal-type-assertion
+        let mock_event = <HTMLInputEvent> {
+            target: {
+                files: null,
+            }
+        };
+
+        expect(() => file_upload_component.add_files_from_button(mock_event)).toThrow(
+            'Files property of event target is unexpectedly null'
+        );
+    });
+
+    test('add_dropped_files throws an error when event.target is null', () => {
+        // tslint:disable-next-line:no-object-literal-type-assertion
+        let mock_event = <DragEvent> {
+            target: null,
+            preventDefault: () => {},
+            stopPropagation: () => {}
+        };
+
+        expect(() => file_upload_component.add_dropped_files(mock_event)).toThrow(
+            'Target is null'
+        );
+    });
+
+    test("Users can delete files after they've been updated", () => {
+        file_upload_component.d_files.push(fake_file_1);
+        file_upload_component.d_files.push(fake_file_2);
+
+        file_upload_component.check_for_emptiness(fake_file_1);
+        file_upload_component.check_for_emptiness(fake_file_2);
+
+        expect(file_upload_component.d_files.length).toEqual(2);
+        expect(file_upload_component.d_empty_filenames.size).toEqual(1);
+
+        let file_delete_buttons = wrapper.findAll('.remove-file-button');
+        file_delete_buttons.at(0).trigger('click');
+
+        expect(file_upload_component.d_files.length).toEqual(1);
+        expect(file_upload_component.d_empty_filenames.has(fake_file_2.name)).toBe(true);
+
+        file_delete_buttons = wrapper.findAll('.remove-file-button');
+        file_delete_buttons.at(0).trigger('click');
+
+        expect(file_upload_component.d_files.length).toEqual(0);
+        expect(file_upload_component.d_empty_filenames.size).toEqual(0);
+    });
+
+    test('Users can successfully submit files when all files are non-empty', () => {
+        let final_submit_button = wrapper.find('.submit-files-button');
+
+        file_upload_component.d_files.push(fake_file_1);
+        file_upload_component.d_files.push(fake_file_3);
+
+        file_upload_component.check_for_emptiness(fake_file_1);
+        file_upload_component.check_for_emptiness(fake_file_3);
+
+        final_submit_button.trigger('click');
+
+        expect(wrapper.emitted().submit_click.length).toBe(1);
+        expect(empty_files_present_modal.is_open).toBe(false);
+    });
+});
+
+describe("File Upload tests concerning the nested modal", () => {
+    let wrapper: Wrapper<FileUpload>;
+    let file_upload_component: FileUpload;
+    let final_submit_button: Wrapper<Vue>;
+    let empty_files_present_modal: Modal;
+    let rows1: string[];
```

Comment:
Inline the value that's assigned to these `rows` variables and then remove the variables